pylint/reporters/base_reporter.py

```diff
@@ -35,7 +35,7 @@ def handle_message(self, msg: Message) -> None:
         """Handle a new message triggered on the current file."""
         self.messages.append(msg)
 
-    def set_output(self, output=None):
+    def set_output(self, output: Optional[TextIO] = None) -> None:
         """set output stream"""
         self.out = output or sys.stdout
```

Comment:
```suggestion
    @property
    def out(self) -> TextIO:
        return self.__out


    @out.setter
    def out(self, output: TextIO = sys.stdout) -> None:
        self.__out = output or sys.stdout
        
    def set_output(self, output: Optional[TextIO] = None) -> None:
        """set output stream"""
        warnings.warn("'set_output' will be removed in 3.0, please use reporter.out =stream instead, DeprecationWarning)
        self.out = output
```

Reply:
I don't really understand why we need the `property` and `setter`. Wouldn't adding a depreciation to `set_output` be enough?

Reply:
You're right, maybe we don't need the ``output or sys.stdout``.

Reply:
Because what is a false stream anyway ? And why would one assign a false stream to ``reporter.out`` ?

Reply:
```suggestion
    def set_output(self, output: TextIO = sys.stdout) -> None:
        """set output stream"""
        warnings.warn("'set_output' will be removed in 3.0, please use 'reporter.out = stream' instead, DeprecationWarning)
        self.out = output
```

Reply:
I don't fully understand your question, but I think this allows extensions to get the output of `pylint` into a different output than `stdout` so they can handle it immediately.
Having this be resolved in the `__init__` is indeed much better, but I wonder why you added the `@property` and `@out.setter`? Keeping `set_output` in with a `DeprecationWarning` seems best to me, as I don't think we fully know how extensions and users might interact with this. You could consider this function to be part of the public API as it is in one of the base classes.

Reply:
Yes, I agree, I added the property in order to handle the case where output is falsey (so sys.stdout is used instead) to not break compatibility. But if someone is using a falsey stream then if it does not work it's on them.

Reply:
See latest commit 😄